pkg/resource/deploy/worker_pool.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"errors"
+	"runtime"
+	"sync"
+)
+
+// A worker pool that can be used to constrain the number of concurrent workers.
+type workerPool struct {
+	numWorkers int
+	// This channel is used as a semaphore to limit how many workers can be active
+	sem chan struct{}
+	wg  sync.WaitGroup
+
+	errorMutex sync.Mutex
+	cancel     context.CancelFunc
+	errors     []error
+}
```

Comment:
bit of an arch/style nit here

the other worker pool that we use in the engine (i.e. the step executor) uses a pool of long-lived goroutines reading off of a shared `chan func()` that represents the work to be done. I have a slight preference for maintaining the same approach here unless there's a compelling reason to use short-lived goroutines (and maybe there is, as it eliminates idle coroutines?).

Reply:
I could certainly bike-shed on this. In general I prefer this approach because of its generality, the step executor worker pool is single purpose for the step executor, so the workers all have common functionality. 

That being said, this could certainly be implemented with an array of worker goroutines which each pull the next task/thunk from a chan it just feels like an extra step, each scheduled task either just gets wrapped in a goroutine, or gets put onto a channel, selected from the channel, then called. 

I'm not convinced that either is _better_, but if you have a goroutine for each conceptual thread, then you have to handle cleanup to ensure that each is closed correctly, which is easy with `context` but is an additional step